dev/io.openliberty.security.jakartasec.4.0.internal/src/io/openliberty/security/jakartasec/identitystore/InMemoryIdentityStoreDefinitionWrapper.java

```diff
@@ -0,0 +1,421 @@
+
+        /*
+         * Evaluate the configuration. The values will be non-null if the setting is NOT
+         * a deferred EL expression. If it is a deferred EL expression, we will dynamically
+         * evaluate it at call time.
+         */
+        priority = evaluatePriority(false);
+        useFor = evaluateUseFor(false);
+
+        // with the annotation from the application, parse out the values
+        int credentialsLength = (inMemoryIdentityStoreDefinition.value() == null) ? 0 : inMemoryIdentityStoreDefinition.value().length;
+        Credentials[] creds = Arrays.copyOf(inMemoryIdentityStoreDefinition.value(), credentialsLength);
+
+        // convert annotation credentials into an easily search-able HashMap with standard Java types
+        credentials = new HashMap<String, CredentialValue>(credentialsLength);
+        for (int i = 0; i < credentialsLength; i++) {
+            CredentialValue credential = getCredential(creds[i].callerName(), creds[i].password(), creds[i].groups());
+            credentials.put(creds[i].callerName(), credential);
+        }
+    }
+
+    private CredentialValue getCredential(String callerName, @Sensitive String password, String[] groups) {
+        String evaluatedPassword = evaluatePassword(password, false);
+        return new CredentialValue(callerName, (evaluatedPassword == null) ? "" : evaluatedPassword, groups);
+    }
+
+    /**
+     * Evaluate and return the password.
+     *
+     * @param immediateOnly If true, only return a non-null value if the setting is either an
+     *                          immediate EL expression or not set by an EL expression. If false, return the
+     *                          value regardless of where it is evaluated.
+     * @return The password or null if immediateOnly==true AND the value is not evaluated
+     *         from a deferred EL expression.
+     */
+    @FFDCIgnore(IllegalArgumentException.class)
+    private String evaluatePassword(@Sensitive String password, boolean immediateOnly) {
+        String evaluatedPassword;
+
+        // evaluate possible EL expressions OR environment variable specification
+        try {
+            evaluatedPassword = elHelper.processString("password", password, immediateOnly, true);
+        } catch (IllegalArgumentException e) {
+            /*
+             * If deferred expression and called during initialization, return null so the expression can be re-evaluated
+             * again later.
+             */
+            if (immediateOnly && ELHelperWrapper.isDeferredExpression(password)) {
+                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                    Tr.debug(tc, "evaluatePassword", "Returning null since password is a deferred expression and this is called on initialization.");
+                }
+                return null;
+            }
+
+            if (TraceComponent.isAnyTracingEnabled() && tc.isWarningEnabled()) {
```

Comment:
If you only want this warning to come out if trace is enabled, why have it be a translated warning message and not just a trace message?  You have this same pattern in another places well.

```suggestion
            if (tc.isWarningEnabled()) {
```
Either this suggestion or remove the if statement completely.

Reply:
Okay, yes, another copy and paste from similar coding in `com.ibm.ws.security.javaeesec`  ... 

I've updated all the occurrences.

I have simply removed the `if` statement in its entirety.  Regarding outputting a translated message (as opposed to a free text one, if that's what your implying the opposite is?), I preferred the translated message, only for consistency with other messages for a similar error. 

Are you suggesting I don't use a translated message at all, just free text in the warning?

Reply:
The question is do you want the message to come out always (since warning is enabled by default and is a message type and not at trace type) or do you want the warning to come out only if they enable trace?  It is a serviceability question.  Do you want customers to have to enable trace (fine, finer, finest) for any class in order to get the warning messages to come out?  With your latest change you are saying you want them to come out like any other warning message and not require the customer to have any sort of trace enabled.  

I am NOT suggesting putting free text in a warning.  Warnings are messages, not trace and require translated text.

Since it is copy and paste from another place, it would be good to get feedback from the security team if that code also should be updated to have the warning messages come out without customers enabling trace on any other class.  We should be consistent.

Reply:
Hi, yes, my intent was (or should be) that the warning messages are always output like any other warning messages, irrespective of what tracing the customer has enabled.  

I am having a code walkthrough with the wider security team on 24/10 and have made a note to discuss the consistency issue you raise.  Thanks!